plutarch-test/src/Plutarch/Test.hs

```diff
@@ -0,0 +1,205 @@
+  see https://github.com/input-output-hk/plutus/issues/4270
+
+-}
+printTermEvaluated :: forall a. ClosedTerm a -> String
+printTermEvaluated p =
+  case evaluateScript (compile p) of
+    Left _ -> printTerm perror
+    Right (_, _, x) -> printScript x
+
+{- | Asserts that the term evaluates successfully with the given trace sequence
+
+  See also: `plutarchDevFlagDescribe`
+-}
+ptraces :: forall (a :: PType). ClosedTerm a -> [Text] -> Expectation
+ptraces p develTraces =
+  case evaluateScript (compile p) of
+    Left _ -> expectationFailure $ "Term failed to evaluate"
+    Right (_, traceLog, _) -> do
+#ifdef Development 
+      traceLog `shouldBe` develTraces
+#else
+      -- Tracing is disabled in non-developed modes, so we should expect an
+      -- empty trace log.
+      let noTraces = const [] develTraces
+      traceLog `shouldBe` noTraces
+#endif
+
+{- | Like `describe`, but determines description from `Development` CPP flag
+
+  Useful to create two sets of othersise identical group of tests that differ
+  only by `Development` flag. This has the effect of creating two sets of golden
+  tests (with different filepaths) for corresponding flag values.
+
+  Typically meant to be used in conjunction with `ptraces`.
+-}
+plutarchDevFlagDescribe :: forall (outers :: [Type]) inner. TestDefM outers inner () -> TestDefM outers inner ()
+
+-- CPP support isn't great in fourmolu.
+{- ORMOLU_DISABLE -}
+plutarchDevFlagDescribe m =
+#ifdef Development 
+  describe "dev=true" m
+#else
+  describe "dev=false" m
+#endif
+{- ORMOLU_ENABLE -}
+
+{- Asserts the term evaluates without success -}
+pfails :: forall (a :: PType). ClosedTerm a -> Expectation
+pfails p = do
+  case evaluateScript (compile p) of
+    Left _ -> pure ()
+    Right _ -> expectationFailure $ "Term succeeded"
+
+{- Whether to run all or a particular golden test
+
+  Typically you want to use `All` -- this produces printTerm and benchmark
+  goldens.
+
+  Occasionally you want `PrintTerm` because you don't care to benchmark that
+  program.
+
+  Use `Bench` to only benchmark the program.
+-}
+data PlutarchGolden
+  = All
+  | Bench
+  | PrintTerm
+  deriving stock (Eq, Show)
+
+hasBenchGolden :: PlutarchGolden -> Bool
+hasBenchGolden = \case
+  PrintTerm -> False
+  _ -> True
+
+hasPrintTermGolden :: PlutarchGolden -> Bool
+hasPrintTermGolden = \case
+  Bench -> False
+  _ -> True
+
+{- Run golden tests on the given Plutarch program -}
+golden :: PlutarchGolden -> ClosedTerm a -> Spec
+golden pg = golden' pg Nothing
+
+-- | Make golden tests for the given Plutarch program.
+golden' :: forall a. PlutarchGolden -> Maybe String -> ClosedTerm a -> Spec
+golden' pg mk p =
+  goldens' pg mk [("0", popaque p)]
+
+{- | Like `golden` but for multiple programs
+
+  Multiple programs use a single golden file. Each output separated from the
+  keyword with a space.
+-}
+goldens :: PlutarchGolden -> [(String, ClosedTerm a)] -> Spec
```

Comment:
Perhaps this should be `ClosedTerm POpaque` for the sake of consistency and type inference? Your choice.

Reply:
The `a` requires ImpredicativeTypes. Keeping it polymorphic is nice, because it avoids popaque in most golden groups, eg:

<img width="420" alt="image" src="https://user-images.githubusercontent.com/3998/154165804-be013b8d-1f75-45e1-b595-b5c4c71b9959.png">

Reply:
Good point.